Fix compiler warnings

tmadlener opened the pull request:
BEGINRELEASENOTES
- Fix compiler warnings, that were uncovered by #153 and described in #170. Fix them in the core classes and also in the generated ones.
- Enforce no new warnings with `Werror` in the CI builds.

ENDRELEASENOTES

Fixes #170 

These changes should fix in the core classes:
- all of the `Weffc++` warnings that are not yet covered by #156
- all of the `Wshadow-[parameter,variable]` ~not yet covered by #158 (mainly `podio::ObjBase`)~ (26.01.2021, copied relevant changes)
- the one `Wnon-virtual-dtor` warning in `ColWriterBase`
- the two `Wcatch-value=` warnings in `unittest.cpp`
- the `Wreorder` warnings in the `ROOTWriter`
- the `Wpedantic` warnings due to unnecessary semi colons

In the generated classes they should fix:
- All `Weffc++` warnings for the `Collection` classes, ~except the ones that are not fixable due to the root constraints (see #170)~(28.01.21)
- All `Weffc++` warnings for the `Obj` classes ~except for the ones covered by #158~ (26.01.2021, copied relevant changes)
- All `Weffc++` warnings about uninitialized members in the `Data` classes and in the components. (26.01.2021)
- All `Wshadow` warnings ~(partially overlapping with #158 due to a slightly different approach)~ (26.01.2021, copied the relevant changes)

~These changes **do not yet fix**~
- ~`Weffc++` warnings arising in the generated components. Here, I think we first have to think about how to best handle these.~
- ~The `Wsign-compare` warnings in the core as well as in the generated classes (see also the discussion in #170). Also this needs a bit more discussion on how to proceed.~ (26.01.2021)

Some simple statistics (via `make | sed -nr 's/.+ \[-(.*)\].*/\1/p' | sort | uniq -c`)

#### gcc statistics
Before
```
      2 Wcatch-value=
   2338 Weffc++
      1 Wnon-virtual-dtor
      2 Wpedantic
      3 Wreorder
    366 Wshadow
     97 Wsign-compare
     37 Wunused-parameter
     18 Wunused-variable
```
After
```
   468 Weffc++
```

~All of the `Wshadow` warnings and about a third of the `Weffc++` warnings are fixed by some of the changes that are present in #158 and #156.~ (26.01.2021) All of the remaining `Weffc++` warnings are fixed by #156 and #154.

#### clang statistics
(probably the easier to parse and more reliable output):
Before
```
     15 Wdeprecated-copy-dtor
     16 Winconsistent-missing-override
     22 Wmismatched-tags
      1 Wnon-virtual-dtor
      1 Woverlength-strings
      1 Wreorder-ctor
     23 Wshadow
    101 Wsign-compare
     37 Wunused-parameter
      1 Wunused-private-field
     18 Wunused-variable
```
After
```
      1 Woverlength-strings
```

These are mostly the same warnings under a slightly different name, only one or two are actually warning about things that gcc didn't warn about yet.


- [x] Check `Wsign-compare` appearance in edm4hep (i.e. user-facing code)
- [x] Fix remainging `Weffc++`
  - [x] Default initialize component and datatype members
  - [x] Disable for `Collection`s 
- [x] Fix warnings that appear in builds using llvm

petricm commented:
@tmadlener you have to still fix the `brew cask install` call in the mac workflow

tmadlener commented:
> @tmadlener you have to still fix the `brew cask install` call in the mac workflow

Opened #172 for that, since this one might need a bit longer for review.

tmadlener commented:
I have implemented the fixes for `Wsign-compare`. In the end this only required very few changes, mainly in the tests and in two places in the `EventStore`.

I have also copied some changes that are currently in #158. I think that can be closed, as all the relevant changes are here now. Some other changes that are present there do not apply here, due to the slightly different approach there (making the members of `podio::ObjectID` unsigned).

With the changes from #154 and #156 only the "unfixable" `Weffc++` in the `Collections` and the uninitialized component and datatype members remain. I could merge those PRs into this one to have this in only one PR, rather than having to deal with it in several different ones.

tmadlener commented:
Now all members in the `Data` classes are default initialized and `Weffc++` is disabled for the copy-constructor of the `Collection`s. This get's rid of all the warnings in the generated code. The unittests now check for `std::is_standard_layout` and `std::is_trivially_copyable` instead of the deprecated `std::is_pod`. I think these should be the two cases we are most interested in.

CI is currently failing because of `-Werror`. I deliberately put this in, so that #170 can be closed when this is merged.

tmadlener commented:
Looking at the CI build outputs, there are a few more warnings that appear in the builds using clang.

tmadlener commented:
I have now also fixed all but one warning in the llvm/clang builds. Additionally, I have significantly reduced the contents of the generated `selection.xml` (see https://github.com/AIDASoft/podio/issues/111#issuecomment-769249615), so that it is now possible to also delete the copy constructor of the generated collections. This in turn means that there is no need for ignoring `Weffc++` for the generated collections any longer, so I have removed that again.

The one warning that is still present for llvm/builds is `Woverlength-strings`, which is triggered by the root generated dictionary: https://github.com/AIDASoft/podio/pull/171/checks?check_run_id=1786088699#step:5:164

I am not entirely sure if and how that could be fixed, but this will only get worse the larger the generated datamodel gets, as basically all generated headers are just dumped into that `DICTPAYLOAD` string. This is basically the only "unfixable" warning left for now.

tmadlener commented:
I have no added `-Wno-overlength-strings` for the root generated dictionary file, in order to suppress it in all builds (not only the CI ones).

tmadlener commented:
I just realized that ubuntus gcc7 and gcc8 are *very* picky in their shadowing warnings: https://github.com/AIDASoft/podio/pull/171/checks?check_run_id=1861872804#step:5:182

They complain about the fact that a local loop variable `data`, shadows a class method `data()`. For me this is a false positive, and I don't see a reason to alter the templates in this case. I have instead renamed the member of the corresponding class from `data` to `component`. This will at least make the CI work, but doesn't really solve the underlying problem. Starting from gcc9 this is no longer flagged, and so also users of podio should not run into similar problems with newer compilers.

petricm commented:
@tmadlener there is a LCG_99 ubuntu2004 gcc9 build

tmadlener commented:
> @tmadlener there is a LCG_99 ubuntu2004 gcc9 build

That is good to know. Thanks. However, at least for now it seems that these older Ubuntu builds are also surfacing some more legitimate unused variable warnings. Unfortunately they also now have one that is definitely a false positive here: https://github.com/AIDASoft/podio/pull/171/checks?check_run_id=1872871018#step:5:94 both of these are used.

The more legitimate case is the following in ROOTReader:
```cpp
std::pair<TTree*, unsigned> getLocalTreeAndEntry(const std::string&);
auto [tree, dummy] = getLocalTreeAndEntry("evt_metadata");
// only use tree here, but not dummy
```
where `dummy` is not used. Unfortunately "ignoring" parts of the structured binding is not (yet) possible in c++. The easiest solution in this case is to simply not do the structured binding, and instead access the pair with `.first` directly.

Alternatively we could switch to newer versions of Ubuntu and gcc. That should hopefully also fix the false positive.

tmadlener commented:
> Unfortunately they also now have one that is definitely a false positive here: https://github.com/AIDASoft/podio/pull/171/checks?check_run_id=1872871018#step:5:94 both of these are used.

Just for completeness, these were actually accurate warnings, but they were attributed to the wrong line in the source file. They were flagging https://github.com/AIDASoft/podio/blob/72e472cd3460c0b486a6514bcedbfdc32f0a2056/src/ROOTWriter.cc#L94
but should instead have flagged L99 in the same file.

Now also these are fixed, and I think this PR is now ready for review and eventual merging.

hegner reviewed (commented):
thanks!
